File: leetcode/3838.py
class Solution:
    def mapWordWeights(self, words: List[str], weights: List[int]) -> str:
        def calculate_weight(word: str) -> int:
            return sum( weights[ord(c) - 97] for c in word ) % 26

        def reverse_lookup(i: int) -> str:
            return chr((25 - i) + 97)

        chars: List[str] = (
            reverse_lookup(calculate_weight(word)) for word in words
        )

        return ''.join(chars)

File: leetcode/3838.kt
class Solution {
    fun mapWordWeights(words: Array<String>, weights: IntArray): String =
        words
            .map { word ->
                val weight =
                    word
                        .toList()
                        .map { weights[it.toIndex()] }
                        .sum()
                reverseAlphabetLookup(weight)
            }.joinToString("")

    fun Char.toIndex(): Int = this.toInt() - 'a'.toInt()

    fun reverseAlphabetLookup(weight: Int): String {
        val index = 25 - weight % 26
        return (index + 'a'.toInt()).toChar().toString()
    }
}
